### PR TITLE
core: refactor `PolarisSecurable` with `PathSegment`

### PR DESCRIPTION
**Pull request by sungwy**
While working on https://github.com/apache/polaris/pull/3999 we realized that `PolarisSecurable` requires an Entity representation with its parent relationships and their corresponding `EntityType`.

This PR introduces a new record  type `PathSegment` that has `name` and `EntityType` that represents a segment of the resource path, that can be ordered to construct a `PolarisSecurable`.

This is not yet a breaking change, because a release hasn't been made since `PolarisSecurable` was introduced.

## Checklist
- [ ] 🛡️ Don't disclose security issues! (contact security@apache.org)
- [x] 🔗 Clearly explained why the changes are needed, or linked related issues: Fixes #
- [ ] 🧪 Added/updated tests with good coverage, or manually tested (and explained how)
- [ ] 💡 Added comments for complex logic
- [ ] 🧾 Updated `CHANGELOG.md` (if needed)
- [ ] 📚 Updated documentation in `site/content/in-dev/unreleased` (if needed)

### CODE REVIEW

File: polaris-core/src/main/java/org/apache/polaris/core/auth/PolarisSecurable.java [hunk 1]
```diff
@@ -18,30 +18,82 @@
...
+  }
+
+  @Value.Check
+  default void validate() {
+    Preconditions.checkState(getPathSegments() != null, "PathSegments must be non-null");
+    Preconditions.checkState(
+        !getPathSegments().isEmpty(), "PathSegments must contain at least one segment");
+    Preconditions.checkState(
+        getPathSegments().get(0).entityType().isTopLevel(),
+        "PathSegments must start with a top-level entity");
```
[the remark this answers is not in the document]
Comment (sungwy) [reply polaris-core/src/main/java/org/apache/polaris/core/auth/PolarisSecurable.java]: thanks Copilot. I'll add null validation in `PathSegment`

File: polaris-core/src/main/java/org/apache/polaris/core/auth/PolarisSecurable.java [hunk 2]
```diff
@@ -18,30 +18,82 @@
...
+      Preconditions.checkState(
+          getParents().isEmpty(),
+          "top-level securable leaf=%s must not declare parents",
+          getLeaf());
+    } else {
+      for (int i = 1; i < getPathSegments().size(); i++) {
+        PathSegment parent = getPathSegments().get(i - 1);
+        PathSegment child = getPathSegments().get(i);
+        Preconditions.checkState(
+            child.entityType().getParentType() == parent.entityType(),
```
[the remark this answers is not in the document]
Comment (sungwy) [reply polaris-core/src/main/java/org/apache/polaris/core/auth/PolarisSecurable.java]: yep - that's a good edge case. Thanks Copilot

File: polaris-core/src/test/java/org/apache/polaris/core/auth/PolarisSecurableTest.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+            new PathSegment(PolarisEntityType.CATALOG, "catalogA"),
+            new PathSegment(PolarisEntityType.NAMESPACE, "ns1"));
+    assertThat(table.getPathSegments())
+        .containsExactly(
+            new PathSegment(PolarisEntityType.CATALOG, "catalogA"),
+            new PathSegment(PolarisEntityType.NAMESPACE, "ns1"),
+            new PathSegment(PolarisEntityType.TABLE_LIKE, "table1"));
+  }
+
+  @Test
```
[the remark this answers is not in the document]
Comment (sungwy) [reply polaris-core/src/test/java/org/apache/polaris/core/auth/PolarisSecurableTest.java:108]: adding more test coverage

File: polaris-core/src/main/java/org/apache/polaris/core/auth/PolarisSecurable.java [hunk 4]
```diff
@@ -18,30 +18,84 @@
...
+  @Nonnull
+  List<PathSegment> getPathSegments();
+
+  /** Returns the leaf segment of the path. */
   @Nonnull
-  PolarisEntityType getEntityType();
+  @Value.Derived
+  default PathSegment getLeaf() {
+    List<PathSegment> pathSegments = getPathSegments();
+    return pathSegments.get(pathSegments.size() - 1);
```
Comment (adutra) [inline polaris-core/src/main/java/org/apache/polaris/core/auth/PolarisSecurable.java:60]:
```suggestion
    return getPathSegments().getLast();
```

File: polaris-core/src/main/java/org/apache/polaris/core/auth/PolarisSecurable.java [hunk 5]
```diff
@@ -18,30 +18,84 @@
...
-  List<String> getNameParts();
+  @Value.Derived
+  default List<PathSegment> getParents() {
+    List<PathSegment> pathSegments = getPathSegments();
+    return pathSegments.subList(0, pathSegments.size() - 1);
+  }
+
+  @Value.Check
+  default void validate() {
+    Preconditions.checkState(getPathSegments() != null, "PathSegments must be non-null");
```
Comment (adutra) [inline polaris-core/src/main/java/org/apache/polaris/core/auth/PolarisSecurable.java]: Already validated by the immutables library:

```suggestion
```

File: polaris-core/src/main/java/org/apache/polaris/core/auth/PolarisSecurable.java [hunk 6]
```diff
@@ -18,30 +18,84 @@
...
+    Preconditions.checkState(
+        !getPathSegments().isEmpty(), "PathSegments must contain at least one segment");
+    Preconditions.checkState(
+        getPathSegments().get(0).entityType().isTopLevel(),
+        "PathSegments must start with a top-level entity");
+    for (PathSegment parent : getParents()) {
+      Preconditions.checkState(
+          parent.entityType() != PolarisEntityType.ROOT,
+          "PathSegments must not include ROOT for securable leaf=%s",
```
Comment (adutra) [inline polaris-core/src/main/java/org/apache/polaris/core/auth/PolarisSecurable.java]: I'm not following, shouldn't the highest ancestor path segments be allowed to be `ROOT` for some securables?

File: polaris-core/src/main/java/org/apache/polaris/core/auth/PolarisSecurable.java [hunk 7]
```diff
@@ -18,30 +18,84 @@
...
+          getLeaf());
+    }
+    if (getLeaf().entityType().isTopLevel()) {
```
Comment (adutra) [inline polaris-core/src/main/java/org/apache/polaris/core/auth/PolarisSecurable.java:84]: I think we are missing unit tests for some of the validation failures.

File: polaris-core/src/test/java/org/apache/polaris/core/auth/AuthorizationRequestTest.java [hunk 8]
```diff
@@ -36,7 +37,9 @@ void hasSecurableTypeReturnsTrueForPrincipalTarget() {
             PolarisAuthorizableOperation.LOAD_TABLE,
             List.of(
                 AuthorizationTargetBinding.of(
-                    PolarisSecurable.of(PolarisEntityType.PRINCIPAL, List.of("alice")), null)));
+                    PolarisSecurable.of(
+                        List.of(new PathSegment(PolarisEntityType.PRINCIPAL, "alice"))),
```
Comment (adutra) [inline polaris-core/src/test/java/org/apache/polaris/core/auth/AuthorizationRequestTest.java]: nit:

```suggestion
                    PolarisSecurable.of(new PathSegment(PolarisEntityType.PRINCIPAL, "alice")),
```

(same for other occurrences in this class)

File: polaris-core/src/test/java/org/apache/polaris/core/auth/AuthorizationRequestTest.java [hunk 9]
```diff
@@ -64,11 +70,18 @@ void hasSecurableTypeReturnsTrueForCatalogRoleAcrossMultipleBindings() {
...
+                        List.of(
+                            new PathSegment(PolarisEntityType.CATALOG, "catalog"),
+                            new PathSegment(PolarisEntityType.NAMESPACE, "ns"))),
                     null),
                 AuthorizationTargetBinding.of(
-                    PolarisSecurable.of(PolarisEntityType.CATALOG, List.of("catalog")),
-                    PolarisSecurable.of(PolarisEntityType.CATALOG_ROLE, List.of("catalog-role")))));
+                    PolarisSecurable.of(
+                        List.of(new PathSegment(PolarisEntityType.CATALOG, "catalog"))),
+                    PolarisSecurable.of(
```
Comment (adutra) [inline polaris-core/src/test/java/org/apache/polaris/core/auth/AuthorizationRequestTest.java]: Given the usage patterns, I think we could have a new method with varargs:


```java
  static PolarisSecurable of(PathSegment first, PathSegment... rest) {
    return ImmutablePolarisSecurable.builder().addPathSegment(first).addPathSegments(rest).build();
  }
```

File: polaris-core/src/main/java/org/apache/polaris/core/auth/PolarisSecurable.java (hunk 6 above)
Comment (sungwy) [reply polaris-core/src/main/java/org/apache/polaris/core/auth/PolarisSecurable.java]: I'm rooting for removing `ROOT` in the parent chain in `PolarisSecurable`, because including `ROOT` as a parent is a strange RBAC specific semantic. I don't think external PolarisAuthorizer implementations like OPA or Ranger would care about a `ROOT` parent.

File: polaris-core/src/main/java/org/apache/polaris/core/auth/PolarisSecurable.java (hunk 7 above)
Comment (sungwy) [reply polaris-core/src/main/java/org/apache/polaris/core/auth/PolarisSecurable.java:84]: Happy to add more coverage

File: polaris-core/src/test/java/org/apache/polaris/core/auth/AuthorizationRequestTest.java (hunk 9 above)
Comment (sungwy) [reply polaris-core/src/test/java/org/apache/polaris/core/auth/AuthorizationRequestTest.java]: I love the suggestion. Thank you @adutra

File: polaris-core/src/main/java/org/apache/polaris/core/auth/PolarisSecurable.java (hunk 5 above)
Comment (sungwy) [reply polaris-core/src/main/java/org/apache/polaris/core/auth/PolarisSecurable.java]: noted

File: polaris-core/src/main/java/org/apache/polaris/core/auth/PolarisSecurable.java (hunk 6 above)
Comment (dimas-b) [reply polaris-core/src/main/java/org/apache/polaris/core/auth/PolarisSecurable.java]: I agree that `ROOT` should not be a factor in authZ decisions in general. Existing native RBAC uses it, which is fine... historical reasons.

It's like appending `.` to DNS names - it makes them formally absolute, but it makes no practical difference in most cases :sweat_smile: 

Let's standardize on starting paths at the catalog in AuthZ SPI.

File: polaris-core/src/main/java/org/apache/polaris/core/auth/PolarisSecurable.java (hunk 4 above)
Comment (sungwy) [reply polaris-core/src/main/java/org/apache/polaris/core/auth/PolarisSecurable.java:60]: Looks like we are still compiling at 17 with `polaris-client`: https://github.com/apache/polaris/blob/f39a9a38c6b6d08b26f2f90713e54808f4062218/build-logic/src/main/kotlin/polaris-client.gradle.kts#L24

Maybe we can keep this PR with the current syntax and raise a DISCUSS thread to assess moving `polaris-client` compilation to 21?

Comment (dimas-b) [reply polaris-core/src/main/java/org/apache/polaris/core/auth/PolarisSecurable.java:60]: Java 21 discussion is on-going (although slowly 😅 ): https://lists.apache.org/thread/v8ghm3k5nh9xp9x0xdo2kknk5ph0rnmb

File: polaris-core/src/main/java/org/apache/polaris/core/auth/PolarisSecurable.java [hunk 10]
```diff
@@ -18,30 +18,87 @@
...
+  List<PathSegment> getPathSegments();
+
+  /** Returns the leaf segment of the path. */
   @Nonnull
-  PolarisEntityType getEntityType();
+  @Value.Derived
+  default PathSegment getLeaf() {
+    List<PathSegment> pathSegments = getPathSegments();
+    Preconditions.checkState(
+        !pathSegments.isEmpty(), "PathSegments must contain at least one segment");
```
Comment (dimas-b) [inline polaris-core/src/main/java/org/apache/polaris/core/auth/PolarisSecurable.java:59]: This check is a bit pointless, given the `validate()` method 🤔

File: polaris-core/src/main/java/org/apache/polaris/core/auth/PolarisSecurable.java [hunk 11]
```diff
@@ -18,30 +18,87 @@
...
+  }
+
+  @Value.Check
+  default void validate() {
+    Preconditions.checkState(
+        !getPathSegments().isEmpty(), "PathSegments must contain at least one segment");
+    Preconditions.checkState(
+        getPathSegments().get(0).entityType().isTopLevel(),
+        "PathSegments must start with a top-level entity");
+    for (PathSegment parent : getParents()) {
```
Comment (dimas-b) [inline polaris-core/src/main/java/org/apache/polaris/core/auth/PolarisSecurable.java]: Why check only parents? I guess `ROOT` is not meaningful as a leaf element either 🤔

File: polaris-core/src/test/java/org/apache/polaris/core/auth/AuthorizationRequestTest.java [hunk 12]
```diff
@@ -36,7 +37,8 @@ void hasSecurableTypeReturnsTrueForPrincipalTarget() {
             PolarisAuthorizableOperation.LOAD_TABLE,
             List.of(
                 AuthorizationTargetBinding.of(
-                    PolarisSecurable.of(PolarisEntityType.PRINCIPAL, List.of("alice")), null)));
+                    PolarisSecurable.of(new PathSegment(PolarisEntityType.PRINCIPAL, "alice")),
```
Comment (dimas-b) [inline polaris-core/src/test/java/org/apache/polaris/core/auth/AuthorizationRequestTest.java:40]: nit: maybe add a convenience `.of()` method for this? It looks like a common pattern.

File: polaris-core/src/main/java/org/apache/polaris/core/auth/PathSegment.java [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.polaris.core.auth;
+
+import com.google.common.base.Preconditions;
+import org.apache.polaris.core.entity.PolarisEntityType;
+
+/** One segment in a fully qualified resource path. */
+public record PathSegment(PolarisEntityType entityType, String name) {
```
Comment (dimas-b) [inline polaris-core/src/main/java/org/apache/polaris/core/auth/PathSegment.java:25]: nit: maybe add a convenience `.pathSegment()` builder method? Callers would be able to statically import it and have less verbose code.

File: polaris-core/src/main/java/org/apache/polaris/core/auth/PolarisSecurable.java (hunk 11 above)
Comment (sungwy) [reply polaris-core/src/main/java/org/apache/polaris/core/auth/PolarisSecurable.java]: yeah I guess we could expand it to the leaf as well

File: polaris-core/src/main/java/org/apache/polaris/core/auth/PolarisSecurable.java (hunk 10 above)
Comment (sungwy) [reply polaris-core/src/main/java/org/apache/polaris/core/auth/PolarisSecurable.java:59]: I agree, but I thought of just keeping this to make the error more descriptive in the future, if our API evolves... it shouldn't ever happen and `PathSegments must contain at least one segment` feels more descriptive than `ArrayIndexOutOfBoundsException`.

File: polaris-core/src/test/java/org/apache/polaris/core/auth/AuthorizationRequestTest.java (hunk 12 above)
Comment (sungwy) [reply polaris-core/src/test/java/org/apache/polaris/core/auth/AuthorizationRequestTest.java:40]: that's a nice neat trick :)

Comment (sungwy) [reply polaris-core/src/test/java/org/apache/polaris/core/auth/AuthorizationRequestTest.java:40]: hmm I thought about this a bit more, and I'm thinking of just keeping the plain constructors for now. I think it's explicit and consistent for securables with single PathSegment and multiple segments

Comment (dimas-b) [reply polaris-core/src/test/java/org/apache/polaris/core/auth/AuthorizationRequestTest.java:40]: ok, no worries

File: polaris-core/src/main/java/org/apache/polaris/core/auth/PolarisSecurable.java (hunk 10 above)
Comment (dimas-b) [reply polaris-core/src/main/java/org/apache/polaris/core/auth/PolarisSecurable.java:59]: well, I do not see how this check can ever fail 🤔 I believe the failure will happen in the `validate()` method and the object will never get constructed for this method to be called on 🤔

**Comment by sungwy** [comment]
Requesting for re-review @dimas-b - I think this would be safe to merge since this is a new, unused class

**Review by dimas-b — APPROVED** [review]
I'm fine with merging now. If @adutra has any follow-up feedback we can always adjust this code later.
